handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java

```diff
@@ -229,4 +250,112 @@ protected void configurePipeline(ChannelHandlerContext ctx, String protocol) thr
             group.shutdownGracefully(0, 0, TimeUnit.MICROSECONDS);
         }
     }
+
+    @Test
+    public void testReplaceHandler() throws Exception {
+        final String sniHost = "sni.netty.io";
+        LocalAddress address = new LocalAddress("test");
+        EventLoopGroup group = new DefaultEventLoopGroup(1);
+        Channel sc = null;
+        Channel cc = null;
+
+        SelfSignedCertificate cert = new SelfSignedCertificate();
+
+        try {
+            final SslContext sslServerContext = SslContextBuilder
+                    .forServer(cert.key(), cert.cert())
+                    .sslProvider(SslProvider.OPENSSL)
+                    .build();
+
+            final Mapping<String, SslContext> mapping = new Mapping<String, SslContext>() {
+                @Override
+                public SslContext map(String input) {
+                    return sslServerContext;
+                }
+            };
+
+            final Promise<Void> releasePromise = group.next().newPromise();
+
+            final SniHandler handler = new SniHandler(mapping) {
+                @Override
+                protected void replaceHandler(ChannelHandlerContext ctx,
+                        String hostname, final SslContext context, SSLEngine engine)
+                        throws Exception {
+
+                    boolean success = false;
+                    try {
+                        // By replacing the SniHandler with the SslHandler we're effectively
```

Comment:
this comment is a bit confusing ... in general this is not true because as you indicated the lifetime of SslContext is generally not tied to an individual `SslHandler`.